Proposal: A more complete app template for Flutter

brianegan opened the pull request:
Fixes #76379

  > Flutter has one starter app template: the Counter app. It gets the user going quickly, and teaches low-level basics in 66 lines of code. We don't, however, provide a template that would be closer to a complete app, with several widgets in separate files, app-wide state, navigation or state restoration.

## Summary

I've worked with @filiph to develop a new template application that includes two features: A static list / detail feature and a dynamic settings feature. When a user changes the Theme in the settings screen, the application responds in kind.

The template takes the following notable approaches. Reasons behind each feature / approach is documented in the ["New Flutter Template: Final Decisions" document.](https://docs.google.com/document/d/1sXAPfPaaaskbzRoA4fbAoZjohcd6jZXYYJJQd0lXcuA/edit?usp=sharing)

  - The template targets Intermediate Flutter developers. "Intermediate Flutter Developer" is defined by the "Flutter Learning Journey's" document.
  - Generally follows the [Simple App State Management guide](https://flutter.dev/docs/development/data-and-backend/state-mgmt/simple), using a ChangeNotifier to coordinate multiple Widgets.
  - Includes an analysis_options.yaml file
  - Generates localizations by default using arb files.
  - Includes an example image and establishes 1x, 2x and 3x folders for image assets.
  - Uses a "feature-first" folder organization
  - Navigator 1.0 with pushNamed for Navigation

## Approach

  1. Spent one week on research and competitive analysis. This involved recruiting and interviewing beginner to intermediate Flutter developers to determine their current learning targets and needs. Furthermore, I reviewed popular architecture and state management libraries from the Flutter community to determine the current advice and best practices offered by the Flutter ecosystem. Finally, I researched templates from "competing" frameworks, such as Android, iOS, Xamarin, React and Angular to see what similar frameworks offer in regards to templates. This research has been compiled into the ["New Flutter Template: Research and Competitive Analysis"](https://docs.google.com/document/d/10qG-NKoDrbozKzHW1JjaUHuKWor563p8JERNhkYlsVQ/edit?usp=sharing) Google doc that informed our subsequent work. Please note: This document includes some notes from the user interviews. It does not contain any personally revealing information, but to ensure privacy, I've restricted access to this document.
  2. Next, I spent 3 weeks coding up various approaches. This resulted in the creation of [~25 different potential templates](https://github.com/brianegan/new_flutter_template/tree/main/evaluations). From these evaluations we learned the different types of decisions we need to make for a final template. For example: What should the template do? Should it include localization? Logging? Responsive Breakpoints? Etc.
  3. We then asked for more [community feedback on these decisions](https://github.com/brianegan/new_flutter_template/issues). We used this feedback + all of the information gathered in our previous steps to make our final decisions. The decisions we've reached are documented in the ["New Flutter Template: Final Decisions" document.](https://docs.google.com/document/d/1sXAPfPaaaskbzRoA4fbAoZjohcd6jZXYYJJQd0lXcuA/edit?usp=sharing)
  4.  Now, we've taken those decisions and distilled them into this template proposal!

## Preview

https://user-images.githubusercontent.com/126604/108721724-a0ad9380-7522-11eb-97b9-44307757d93e.mp4


https://user-images.githubusercontent.com/126604/109329104-7d545280-785a-11eb-863b-561e7ac8c95d.mp4



## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

brianegan committed:
First version of a proposed new app template

brianegan committed:
Update Template Manifest

filiph commented:
Hi @Hixie, would you please have a look at this PR, and/or suggest other reviewers?

This is the project that adds a more complete template (`flutter create -t full`, go/flutter-more-complete-template). Issue at #76379.

We have a [24-page document](https://docs.google.com/document/d/1sXAPfPaaaskbzRoA4fbAoZjohcd6jZXYYJJQd0lXcuA/edit?usp=sharing) explaining all the micro-decisions that went into this. I recommend skimming through it since it may pre-empt some understandable questions and raised eyebrows.

zanderso commented:
Sorry if this is in the linked docs, but what is the plan for testing this?

Relatedly, I'm wondering if this could be an opportunity for demonstrating testing practices by including a few more non-trivial tests in the template itself.

filiph commented:
> Sorry if this is in the linked docs, but what is the plan for testing this?

Is there a test for the current `flutter create` template (the "counter app") that we could look at for inspiration?

> Relatedly, I'm wondering if this could be an opportunity for demonstrating testing practices by including a few more non-trivial tests in the template itself.

[Here](https://docs.google.com/document/d/1sXAPfPaaaskbzRoA4fbAoZjohcd6jZXYYJJQd0lXcuA/edit#heading=h.2uh3wmm1ck9o)'s our thinking on this. In short: we see that, with the current template, developers introduce breaking changes to the template in the first 5 minutes of their work, and then the smoke test is broken (often statically broken, so there are analyzer errors and squiggly lines). The effect is that the (otherwise very educational!) `test/widget_test.dart` file gets deleted or commented out. For that reason, the approach is to have relatively trivial tests that _can't_ be broken (they don't depend on `lib/`) but that still provide some boilerplate (`test('...', () {})`) and links to documentation.

Nothing here is set in stone, though.

jonahwilliams commented:
We need to find a way to not duplicate the existing platform files. We occasionally have to update these, and having 2 copies of everything is going to increase the chances that we accidentally get out of sync somewhere.

zanderso commented:
`flutter create` tests live in https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/commands.shard/permeable/create_test.dart. We currently generate the app, analyze it, check formatting, etc., but we don't have devicelab tests that run it, AFAIK.

Hixie commented on packages/flutter_tools/templates/list_detail_app/analysis_options.yaml.tmpl:
```diff
@@ -0,0 +1,16 @@
+linter:
+  rules:
+    prefer_void_to_null: true
+    always_specify_types: false
```
we shouldn't have an analysis_options file here. it should just default to the current default.

Hixie commented on packages/flutter_tools/templates/list_detail_app/lib/main.dart.tmpl:
```diff
@@ -0,0 +1,19 @@
+import 'package:flutter/material.dart';
+import 'package:{{projectName}}/src/app.dart';
+import 'package:{{projectName}}/src/settings/settings_controller.dart';
+import 'package:{{projectName}}/src/settings/settings_service.dart';
```
generally we use relative imports within a project, for many reasons, but one reason that should be sufficient here is just that it would prevent the project name from getting hardcoded into the project source, which is a huge pain.

Hixie commented on packages/flutter_tools/templates/list_detail_app/lib/src/app.dart.tmpl:
```diff
@@ -0,0 +1,69 @@
+    // Whenever the user updates their settings, the MaterialApp is rebuilt.
+    return AnimatedBuilder(
+      animation: settingsController,
+      builder: (context, _) {
```
Please give full types and names for every argument to every closure, that way it's self-documenting. Otherwise, a developer looking at this will have no idea what the second argument is nor what the type of the first one is.

Hixie commented on packages/flutter_tools/templates/list_detail_app/lib/src/app.dart.tmpl:
```diff
@@ -0,0 +1,69 @@
+          theme: ThemeData(),
+          darkTheme: ThemeData.dark(),
+          themeMode: settingsController.themeMode,
+          debugShowCheckedModeBanner: false,
```
please don't disable the debug banner by default
the banner exists to prevent people from accidentally thinking they are in release builds

Hixie commented on packages/flutter_tools/templates/list_detail_app/lib/src/dummy_feature/dummy_item_list_view.dart.tmpl:
```diff
@@ -0,0 +1,57 @@
+      // In contrast to the default ListView constructor, which requires
+      // building all Widgets up front, the ListView.builder constructor lazily
+      // builds Widgets as they’re scrolled into view.
+      body: ListView.builder(
```
would it make sense to use the ListView builder that gives the items a fixed height using the prototype widget?

brianegan replied:
Sorry, I'm not quite sure I understand. Are you suggesting we add an `itemExtent` here to give items a fixed height and convert each item into `Placeholder` widget?

Hixie replied:
I meant using SliverPrototypeExtentList. Filed https://github.com/flutter/flutter/issues/76719 because we really should make that easier.

brianegan replied:
Ah, wow -- I'd never used that one myself! Thanks for the info :)

My question: Since this is aimed at intermediate Flutter developers, do you think it might be best to point them to the simpler `ListView.builder` for now until #76719 is addressed? 

The reason I introduced `ListView.builder` instead of a regular `ListView` was because I was actually asked about the difference during one of my user interviews and I thought it would be nice to highlight. If we use `SliverPrototypeExtentList`, we need to introduce the concepts of `CustomScrollView`, `slivers`, and `SliverChildBuilderDelegate` as well. Overall, it's doable, but might be a bit much given the target audience.

What do you think?

```dart
      // Updated version using SliverPrototypeExtentList
      CustomScrollView(
        slivers: [
          SliverPrototypeExtentList(
            delegate: SliverChildBuilderDelegate(
                  (context, index) {
                final item = items[index];

                return ListTile(
                  title: Text('DummyItem ${item.id}'),
                  leading: const CircleAvatar(
                    // Display the Flutter Logo image asset.
                    foregroundImage:
                    AssetImage('assets/images/flutter_logo.png'),
                  ),
                  onTap: () {
                    Navigator.restorablePushNamed(
                      context,
                      DummyItemDetailsView.routeName,
                    );
                  },
                );
              },
              childCount: items.length,
            ),
            prototypeItem: ListTile(),
          )
        ],
      )
```

Hixie replied:
ListView.builder is probably fine. We can revisit when #76719 is done.

Hixie commented on packages/flutter_tools/templates/list_detail_app/lib/src/settings/settings_controller.dart.tmpl:
```diff
@@ -0,0 +1,44 @@
+///
+/// Controllers glue Data Services to Flutter Widgets. The SettingsController
+/// uses the SettingsService to store and retrieve user settings.
+class SettingsController with ChangeNotifier {
```
why `with` vs `extends`?

brianegan replied:
I actually don't have a strong preference here. 

I went with a mixin since it's a bit more "flexible" in that end-users could extend from another class of their choice if need be. 

The [documentation of the class](https://api.flutter.dev/flutter/foundation/ChangeNotifier-class.html) simply reads "A class that can be extended or mixed in," but is extending the recommended approach?

Hixie replied:
`with` is fine, I was just curious.

Hixie commented on packages/flutter_tools/templates/list_detail_app/lib/src/settings/settings_view.dart.tmpl:
```diff
@@ -0,0 +1,50 @@
+      ),
+    );
+  }
+}
```
no restoration? showing how to do restoration well seems very valuable for a best practices template

brianegan replied:
Are user settings generally a good candidate for State Restoration? In my experience, it's better to use a proper persistence layer backed up by a local database (shared_preferences, sqflite, hive, etc) or a remote source (REST, GraphQL, etc).

That way the user's preferred theme is properly persisted when the user closes the application or restarts the operating system.

In my view, State Restoration is more appropriate for things like Navigation Stack or Forms that need to be restored when the user puts an app into the background then wants to come back to the app.

What do you think?

Hixie replied:
i agree it wouldn't make sense for this particular setting, but I meant for the routes themselves. when the user is on the settings page, then switches to another app, then comes back, without restoration, there's a risk they'll end up on the main page again.

brianegan replied:
Good call. I've added Route restoration!

Hixie commented:
We should definitely have tests for this app since it's non-trivial. I agree they shouldn't be in the template.
I think it might be worth branding this as a "best practices" template rather than "full" or "complete" or similar.
Thanks for the decision doc, that was very helpful. I think the analysis_options decision should be revisited in light of the comment I added to the doc. There's a large value in not having a file in the template there because everything in the template gets left behind when we update.
I concur with @jonahwilliams that we should make sure to not duplicate the platform-specific bits.

brianegan committed:
Updates based on feedback:

  - Relative imports
  - Provide full types to closures
  - Enable debugShowCheckedModeBanner

brianegan commented:
Thanks to everyone for the reviews! I'm working on addressing all feedback.

  > We should definitely have tests for this app since it's non-trivial.

Could you please describe in a bit more detail what kind of tests you'd like to see? I was going to replicate the tests for the Counter app, but I can include additional Widget and/or Unit tests to cover the functionality of the app as well.

Thanks again :)

Hixie commented:
> Could you please describe in a bit more detail what kind of tests you'd like to see?

Well at a minimum we should verify that we can navigate to each page, and that changing the theme settings works. Might be good to have some golden tests too just in case.

brianegan committed:
Add Route Restoration to new template

brianegan committed:
Remove analysis options from template

brianegan committed:
Shared Platform Files for Application Templates

  - Introduces concept of Merged Templates
  - app_shared contains files common to all application templates
  - app is now a merged template
  - list_detail_app now a merged template
  - add list_detail_app to CLI

brianegan committed:
Add .img.tmpl placeholders to list/detail template

goderbauer commented on packages/flutter_tools/templates/list_detail_app/lib/src/app.dart.tmpl:
```diff
@@ -0,0 +1,73 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_gen/gen_l10n/app_localizations.dart';
+import 'package:{{projectName}}/src/dummy_feature/dummy_item_details_view.dart';
+import 'package:{{projectName}}/src/dummy_feature/dummy_item_list_view.dart';
+import 'package:{{projectName}}/src/settings/settings_controller.dart';
+import 'package:{{projectName}}/src/settings/settings_view.dart';
```
These imports should probably be relative since they reference files form the same library?

goderbauer commented on packages/flutter_tools/templates/list_detail_app/lib/src/app.dart.tmpl:
```diff
@@ -0,0 +1,73 @@
+          // depending on the user's locale.
+          localizationsDelegates: [AppLocalizations.delegate],
+
+          // Use AppLocalizations to configure a the correct application title
```
typo: "a the"

goderbauer commented on packages/flutter_tools/templates/list_detail_app/lib/src/app.dart.tmpl:
```diff
@@ -0,0 +1,73 @@
+          onGenerateTitle: (context) => AppLocalizations.of(context).appTitle,
+
+          // Define a light and dark color theme. Then, read the user's
+          // preferred ThemeMode (light, dark or system default) from the
```
nit:

```suggestion
          // preferred ThemeMode (light, dark, or system default) from the
```

brianegan replied:
Haha, I thought someone might make an Oxford-comma call out! I don't know if this sentence is unclear without it, but added it nonetheless :)

goderbauer commented on packages/flutter_tools/templates/list_detail_app/lib/src/settings/settings_service.dart.tmpl:
```diff
@@ -0,0 +1,14 @@
+  Future<ThemeMode> themeMode() async => ThemeMode.system;
+
+  /// Persists the user's preferred ThemeMode to local or remote storage.
+  Future<void> updateThemeMode(ThemeMode theme) async {}
```
nit: maybe include a comment in the (currently empty) body of this method outlining would you could do here?

goderbauer replied:
(similar to what the class level comment already says, just to make it clear that this method is not supposed to be empty in a production app)

brianegan committed:
Merge templates with test files

brianegan committed:
Use relative imports for app.dart

brianegan commented:
Thanks again for the review, everyone. I've gone through and updated the tool a bit to support the new template. Features include:

  1. Shared Platform Files. To share files, I've introduced the concept of a "merged" template, which combines several source templates into a single destination.
    - `templates/app_shared`  - Contains all files shared between different apps
    - `app` - only files needed by counter app
    - `list_detail_app` - only files needed by the new template.
    -  [Submitted PR to flutter_template_images](https://github.com/flutter/packages/pull/295) to ensure the directory structure is in sync
  2. Added tests
    - We've decided NOT to include any implementation tests in the final template, but we want to test the app before we ship it.
    - To support this Use Case, I've added an `implentation-tests` flag to the `flutter create` command. If this flag is set to true, the template will render `.test.tmpl` files found in the source directories and copy `.golden.tmpl` files from the `flutter_template_images` repo.

Please note: some tests will not pass until we merge https://github.com/flutter/packages/pull/295. 

This is my first time modifying the CLI, so I'm not sure if this is the correct approach! Happy to change it up if need be.

jonahwilliams commented on packages/flutter_tools/lib/src/commands/create_base.dart:
```diff
@@ -110,6 +110,13 @@ abstract class CreateBase extends FlutterCommand {
+      'implementation-tests',
+      help:
+          'Include implementation tests that verify the template functions correctly'
+          'This is normally enabled to test the tool itself.',
```
nit, I would use the same phrasing as above "This is only intended to enable testing of the tool itself."

Hixie replied:
this string contains the substring 'correctlyThis', you probably want to end the first line with a period and a space.

Hixie commented on packages/flutter_tools/lib/src/flutter_project_metadata.dart:
```diff
@@ -15,6 +15,8 @@ enum FlutterProjectType {
   /// It is different than the "module" template in that it exposes and doesn't
   /// manage the platform code.
   app,
+  /// A List/Detail app template that follows community best practices
```
docs/comments should end with a period

Hixie commented on packages/flutter_tools/lib/src/template.dart:
```diff
@@ -26,13 +27,35 @@ import 'dart/package_map.dart';
+/// Furthermore, templates may contain test files and golden images that are only
+/// included when the --include-tests flag is enabled. Test files must end in
+/// `.test.tmpl`. Golden images must end in '.golden.tmpl'. Golden images are
+/// copied from the flutter_template_images repository.
```
the images themselves can't be landed in this repo. cc @Piinks who may be able to show you how to use skia gold to do this.

Piinks replied:
Are the golden file tests being run here in our CI? Or are these tests supposed to be run by users locally after creating their own project?

brianegan replied:
Hi @Piinks :) 

Primary Goal: The Golden Files were added to verify this new template functions correctly on the CI before shipping to users. 

Secondary Goal: During the development of the template, some users (as well as @zanderso above!) asked for more comprehensive test coverage in the template as an educational tool. Therefore, allowing users to generate a template with some Golden files was only meant to be a "bonus" for those curious.

I've taken a look at Flutter Gold, but it appears it is purpose-built for `package:flutter`, whereas this PR modifies `package:flutter_tools`. Is there any documentation on how to adapt that process for Flutter tools?

Piinks replied:
I see! Just to clarify, is the expectation that the same golden file tests will run one way here on our CI and another way when the user creates their own app locally? You are right, flutter tools is not set up to use Gold right now, but I think it could be done. The test config would need to set up a special file comparator that will talk to gold for CI, but that would not work locally.

I am wondering, if the same tests are meant to execute differently this way, and we have a special config in the template to run on our CI in flutter/flutter, will that be confusing to users that see that as an example and think it is how they should set up their tests? 

Or, would it be possible to create a test in the flutter package that creates this app and keeps a golden file of it, and then the template itself can just ship a set of golden file tests that work as the user would expect them? Instead of including golden files here, you could have `flutter test --update-goldens` run when the template is created and it will generate the golden files for users to see and test against. That way we have tests covering this for verifying the template, but users looking to see examples of how to set up these tests don't have to discern between special configs that are just for our testing purposes and what is meant for them.

brianegan replied:
> I see! Just to clarify, is the expectation that the same golden file tests will run one way here on our CI and another way when the user creates their own app locally?

Nope! With the current implementation, you can run `flutter create -t list_detail_app my_app` to get an app with some dummy tests, or run `flutter create -t list_detail_app my_app --implementation-tests` to include widget tests that verify the functionality of the app.

My naive idea was that the process could work the exact same way for the CI or end-users.

  > I am wondering, if the same tests are meant to execute differently this way, and we have a special config in the template to run on our CI in flutter/flutter, will that be confusing to users that see that as an example and think it is how they should set up their tests?

Agreed -- that will be super confusing. Given the fact that we need to use Flutter Goldens at this point instead of one consistent process, I'm think it makes sense to only focus on the main goal of testing this app as part of the Flutter CI workflow.

  > Or, would it be possible to create a test in the flutter package that creates this app and keeps a golden file of it (in Gold), and then the template itself can just ship a set of golden file tests that work as the user would expect them?

I can take a deeper look at that option, but my initial gut reaction is that might be a tough sell, since it would require the the `flutter` package to depend on the `flutter_tools` package. What do you think?

May I ask another question: How strongly do folks feel about golden tests for this app? Would it be a bit simpler and just as effective to add more assertions to the suite that verify a new `ThemeMode` is provided to the `MaterialApp` after a user switches Themes from the Settings page?

brianegan replied:
Thanks  for  your help, @Piinks! After playing around with a couple of options, we've decided to go with the simpler route of using plain-old Widget tests for now :)

jonahwilliams commented on packages/flutter_tools/lib/src/template.dart:
```diff
@@ -26,13 +27,35 @@ import 'dart/package_map.dart';
 /// but without template expansion (data files, etc.), the '.copy.tmpl'
 /// extension may be used.
 ///
+/// Furthermore, templates may contain test files and golden images that are only
```
Sorry I haven't been following the discussion that closely - is the implication here that we're going to golden test the list app? If so we should be using skia gold and not the template images, so that we can handle rolling minor breaking changes from the engine

brianegan replied:
> is the implication here that we're going to golden test the list app

Yes, I was asked to add Golden Tests for the list app. Thanks for the heads up, I'll try to figure out how to use Skia Gold instead.

Hixie commented on packages/flutter_tools/lib/src/template.dart:
```diff
@@ -43,20 +66,23 @@ class Template {
 
-    if (!templateSource.existsSync()) {
+    if (!templateSources.every((Directory directory) => directory.existsSync())) {
       return;
```
is this a silent failure? or? i'm not sure what this if statement is for. when would we go down this branch? might be good to add a comment giving the answer

brianegan replied:
I agree this should probably be a `throwToolExit` if a template source directory does not exist.

@jonahwilliams Do you happen to know why this code branch does not throw an exception? It didn't before, so I assumed that silent return handled some use-case I hadn't discovered.

jonahwilliams replied:
throwToolExit SGTM. The exists sync check was added in 2016, so we might be past that now...

brianegan replied:
Thanks! I've added a throwToolExit and corresponding test.

Hixie commented on packages/flutter_tools/templates/list_detail_app/test/implementation_test.dart.test.tmpl:
```diff
@@ -0,0 +1,71 @@
+import 'package:flutter/material.dart';
```
looks like your .golden.tmpl files are empty, btw.

(in any case, please don't commit binary files to this repo.)

brianegan replied:
Thanks -- that was intentional for the current setup, they were meant to be placeholder files that are replaced by binaries from the template_image_assets repo. Sounds like this will change now that I'm aware of Flutter Gold :)

brianegan committed:
Fix CLI documentation and code comments

brianegan committed:
Remove Golden files from List/Detail Template

brianegan commented:
Hey everyone -- I've updated this PR and [my corresponding PR for the `flutter_template_images` package](https://github.com/flutter/packages/pull/295) to remove the Golden Test files and replace them with normal Widget assertions that verify when a user changes the ThemeMode, the MaterialApp is rebuilt the selected mode. 

Overall, I believe that will achieve the same goal without the added complication of setting up Flutter Goldens for `package:flutter_tools` or making `flutter_tools` a `dev_dependency` of the Flutter package.

Please let me know what you think of this approach!

goderbauer commented on packages/flutter_tools/templates/list_detail_app/lib/src/app.dart.tmpl:
```diff
@@ -0,0 +1,74 @@
+        return MaterialApp(
+          // Providing a restorationScopeId allows the Navigator built by the
+          // MaterialApp to restore the navigation stack when a user returns to
+          // the app.
```
.. after it has been killed while running in the background?

goderbauer commented on packages/flutter_tools/templates/list_detail_app/lib/src/app.dart.tmpl:
```diff
@@ -0,0 +1,74 @@
+          //
+          // The appTitle is defined in .arb files found in the localization
+          // directory.
+          onGenerateTitle: (context) => AppLocalizations.of(context).appTitle,
```
For consistency (see line 26): 

```suggestion
          onGenerateTitle: (BuildContext context) => AppLocalizations.of(context).appTitle,
```

goderbauer commented on packages/flutter_tools/templates/list_detail_app/lib/src/app.dart.tmpl:
```diff
@@ -0,0 +1,74 @@
+          onGenerateRoute: (RouteSettings routeSettings) {
+            return MaterialPageRoute<void>(
+              settings: routeSettings,
+              builder: (context) {
```
`BuildContext context`

goderbauer commented on packages/flutter_tools/templates/list_detail_app/lib/src/dummy_feature/dummy_item_list_view.dart.tmpl:
```diff
@@ -0,0 +1,65 @@
+            ),
+            onTap: () {
+              // Navigate to the details page. If the user leaves and returns to
+              // the app, the navigation stack is restored.
```
..returns to the app after it has been killed while running in the background...

goderbauer commented on packages/flutter_tools/templates/list_detail_app/lib/src/settings/settings_controller.dart.tmpl:
```diff
@@ -0,0 +1,45 @@
+
+  /// Update and persist the ThemeMode based on the user's selection.
+  Future<void> updateThemeMode(ThemeMode newTheme) async {
+    _themeMode = newTheme;
```
technically, we only need to persist and inform if `newTheme != _themeMode`?

brianegan replied:
For sure. I wasn't sure if it made sense to add performance optimizations in at this point, but we could return early from this method if the `newTheme` is equal to the old theme. What do you think?

goderbauer replied:
I think I would include this here with a quick comment explaining the rational.

brianegan replied:
Sure, no problem. I've added it in!

goderbauer commented on packages/flutter_tools/templates/list_detail_app/test/implementation_test.dart.test.tmpl:
```diff
@@ -0,0 +1,73 @@
+}
+
+Finder findApp(ThemeMode themeMode) => find.byWidgetPredicate(
+    (widget) => widget is MaterialApp && widget.themeMode == themeMode);
```
This formatting is a little strange, maybe:

```suggestion
    (widget) => widget is MaterialApp && widget.themeMode == themeMode,
);
```

goderbauer commented on packages/flutter_tools/templates/list_detail_app/test/implementation_test.dart.test.tmpl:
```diff
@@ -0,0 +1,73 @@
+
+Finder findApp(ThemeMode themeMode) => find.byWidgetPredicate(
+    (widget) => widget is MaterialApp && widget.themeMode == themeMode);
+
```
nit: remove empty line?

goderbauer commented on packages/flutter_tools/templates/list_detail_app/test/implementation_test.dart.test.tmpl:
```diff
@@ -0,0 +1,73 @@
+import 'package:flutter/material.dart';
```
Do you still need all the `*.png.golden.tmpl` files? What are those used for now?

brianegan replied:
Oops, nope! Thanks.

goderbauer commented on packages/flutter_tools/templates/list_detail_app/lib/src/dummy_feature/dummy_item_list_view.dart.tmpl:
```diff
@@ -0,0 +1,65 @@
+      // In contrast to the default ListView constructor, which requires
+      // building all Widgets up front, the ListView.builder constructor lazily
+      // builds Widgets as they’re scrolled into view.
+      body: ListView.builder(
```
The listview also needs a restoration id for full state restoration of scroll position

brianegan replied:
Thanks, added the following to the `ListView.builder`:

```dart
        // Providing a restorationId allows the ListView to restore the
        // scroll position when a user leaves and returns to the app after it
        // has been killed while running in the background.
        restorationId: 'dummyItemListView',
```

goderbauer reviewed (commented):
+1 to using regular widget tests over golden tests

The template itself looks pretty good. Haven't looked at the flutter_tools changes that closely, leaving that for @jonahwilliams.

goderbauer commented:
Looks like there are some test failures...

esDotDev commented:
This new example looks amazing! That decision doc was a great read, and the 25 spikes that led to this would be a great learning resource for the community. 

It's probably too late, but the one thing that would really complete this, it is to show an example of arg passing.  That feels like the one obvious gap from a developer POV. It feels odd that the list handler is not passing some sort of id to the detail view, and that the router generator is also not looking for it.

brianegan committed:
Update packages/flutter_tools/templates/list_detail_app/lib/src/app.dart.tmpl

Co-authored-by: Michael Goderbauer <goderbauer@google.com>

brianegan committed:
Update packages/flutter_tools/templates/list_detail_app/test/implementation_test.dart.test.tmpl

Co-authored-by: Michael Goderbauer <goderbauer@google.com>

brianegan committed:
Update app template based on feedback

  - Better restoration comments
  - Types for Closures
  - State restoration for ListView
  - Formatting

brianegan commented:
> Looks like there are some test failures...

Thanks! In order for all tests to pass we need to merge and publish the [corresponding template images PR](https://github.com/flutter/packages/pull/295). I've updated that PR based on the feedback :)

   > This new example looks amazing! That decision doc was a great read, and the 25 spikes that led to this would be a great learning resource for the community.

Yay! Glad to hear it :)

  > It's probably too late, but the one thing that would really complete this, it is to show an example of arg passing. That feels like the one obvious gap from a developer POV. It feels odd that the list handler is not passing some sort of id to the detail view, and that the route generator is also not looking for it.

That was actually a really tricky call and we went back and forth on our final decision :) 

In the end, to handle all of the use cases we wanted to (responsive design, web urls, etc), it felt like we had three options:

  1. No argument passing (what we've got here)
  2. "Hacky" argument passing -- short and sweet, but easy to break and doesn't *really* demonstrate how it should work
  3. Full-featured argument passing -- Maybe a bit too much code for a template, much of it felt like it would need to be deleted or heavily modified by the end-user. Seemed better as a demonstration than a template.

My personal opinion: Navigator 2.0 is a bit hard to use, but the team is now working on making the APIs simpler to use. Once that work has completed, this template should be updated to use it and include Argument passing!

brianegan committed:
Remove outdated comment for renderMerged method

esDotDev commented:
Thanks Brian, that sounds great, a lightweight Router version (when it is possible), would be the cherry on top for sure.

A bit off topic, but I do think this entire journey would make a great post and resource for the community. I found the decision doc super interesting, I'm sure many other devs would too, there's tons of valuable insights in there for devs. 

Also the work that went into it, all the different examples, would be super helpful for people to compare and contrast some state management approaches. This app has just enough complexity that it can probably serve as a great way to "taste" what each SM style is like.

brianegan committed:
throwToolExit if source directory is missing

brianegan committed:
Exit early from updateThemeMode if old and new ThemeMode are identical

brianegan commented:
Thanks again for all the reviews! I've updated the PR to address the previous round of feedback:

  - `throwToolExit` if a template source directory does not exist (plus testing for that)
  - Return early from the `updateNewTheme` method if the new `ThemeMode` is the same as the previous `ThemeMode`

At this point, I think I've addressed all outstanding feedback, but I'm in a bit of a catch 22! The tests on this PR will not pass until we merge and publish the [corresponding template images PR](https://github.com/flutter/packages/pull/295). However, the corresponding images PR cannot be merged until this PR is approved :) 

If this PR looks good at this stage, could a  Flutter team-member please help me resolve this problem by approving both PRs, merging and publishing the template images PR, then letting me know? After that, I can update this PR to point to version 1.0.2 and the new tests I've written for this template will pass.

zanderso commented on packages/flutter_tools/lib/src/flutter_project_metadata.dart:
```diff
@@ -15,6 +15,8 @@ enum FlutterProjectType {
   /// manage the platform code.
   app,
+  /// A List/Detail app template that follows community best practices.
+  list_detail_app,
```
nit: Dart style would be `listDetailApp`

brianegan replied:
Sure, I can understand that is a bit different from the usual Dart Style. However, this enum appears to be the "type safe" name for the template.

Right now, to create the new template, you run the following command:

```
flutter create -t list_detail_app my_app
```

If we change this up based on your suggestion, the CLI would change to:

```
flutter create -t listDetailApp my_app
```

I don't have a strong preference either way, but wanted to highlight the CLI change and verify that camelCase is the correct case to use there.

zanderso replied:
Maybe @jonahwilliams can clarify, but I don't understand why the string provided on the command line has to exactly match the identifier in Dart code. Something a bit too clever is probably happening somewhere if that is the case.

brianegan replied:
With the rename from `list_detail_app` to `skeleton` this is no longer a cause of concern! This now reads `skeleton` instead, conforming to the standard conventions.

Thanks for pointing it out :)

zanderso reviewed (commented):
One style it, and it also looks like there are some test failures that are not related to the missing image.

goderbauer reviewed (commented):
The template LGTM

brianegan committed:
Add list_detail_app to usage tests and fix test filesystem

brianegan commented:
> The template LGTM

Yay! Thanks @goderbauer  -- appreciate your help with this one :)

  > You'll need to update some of the tests:

Thanks @jonahwilliams -- fixed those up!

brianegan committed:
Update DeferredComponentsPrebuildValidator tests

brianegan committed:
Fix Create Test

  - Bump flutter_template_images to 2.0.0
  - Fix typo in variable name

filiph commented:
I was talking to @csells yesterday, and he had some good ideas on alternative naming:

* skeleton
* scaffold
* bootstrap

I really like these. They are close to the core idea of this new template: it's a _skeleton_ on which you can build your app. Nothing more, nothing less.

Picking a name like that would also make the template more accessible to CLI users: `flutter create -t scaffold` rolls better than `flutter create -t list_detail_app`.

brianegan commented:
Those are definitely better. I don't have a strong preference as to which
is best -- please let me know which one you prefer and we can roll with
that :)

On Wed, Mar 10, 2021 at 7:26 PM Filip Hracek ***@***.***>
wrote:

> I was talking to @csells <https://github.com/csells> yesterday, and he
> had some good ideas on alternative naming:
>
>    - skeleton
>    - scaffold
>    - bootstrap
>
> I really like these. They are close to the core idea of this new template:
> it's a *skeleton* on which you can build your app. Nothing more, nothing
> less.
>
> Picking a name like that would also make the template more accessible to
> CLI users: flutter create -t scaffold rolls better than flutter create -t
> list_detail_app.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/flutter/pull/76519#issuecomment-795868568>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA65DCDTURBD7PZCFK52PDTC6TVDANCNFSM4YAM3XIA>
> .
>

filiph commented:
I'm not gonna lie: I really like `skeleton`. I think it's unique, rememberable, and fits the purpose well. It's also an ESL-friendly word (i.e. when learning English, I definitely knew the word "skeleton" before knowing "scaffold" or "bootstrap"). On the other hand, I can imagine people not liking the image that `skeleton` projects, or thinking it's too childish.

@Hixie, @jonahwilliams, @Piinks, @zanderso, @goderbauer and others: What do you think about `flutter create -t skeleton`?

My second choice would be `flutter create -t scaffold`.

brianegan commented:
One more thing to note: If we decide to change name at this point, we'll
need to publish another version of flutter_template_images as well (since
we just published a version that uses the "list_detail_app" naming).




On Wed, Mar 10, 2021 at 8:57 PM Filip Hracek ***@***.***>
wrote:

> I'm not gonna lie: I really like skeleton. I think it's unique,
> rememberable, and fits the purpose well. It's also an ESL-friendly word
> (i.e. when learning English, I definitely knew the word "skeleton" before
> knowing "scaffold" or "bootstrap"). On the other hand, I can imagine people
> not liking the image that skeleton projects, or thinking it's too
> childish.
>
> @Hixie <https://github.com/Hixie>, @jonahwilliams
> <https://github.com/jonahwilliams>, @Piinks <https://github.com/Piinks>,
> @zanderso <https://github.com/zanderso>, @goderbauer
> <https://github.com/goderbauer> and others: What do you think about flutter
> create -t skeleton?
>
> My second choice would be flutter create -t scaffold.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/flutter/pull/76519#issuecomment-795999381>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA65DCFVGQE3EWTG5B2RATTC66I7ANCNFSM4YAM3XIA>
> .
>

brianegan committed:
Use consistent Filesystem for Tests

brianegan commented:
Hey everyone -- we've merged up and deployed the flutter_template_images package,  which means the tests are now passing for this PR :)

Please let me know if there are any more code changes or template changes you'd like to see!

jonahwilliams commented:
I don't have any particular opinions on the name

Hixie commented:
I'd avoid `scaffold` because of the Scaffold widget. `bootstrap` is a little idiomatic (and may be confused with the JS framework?). `skeleton` seems fine to me.

brianegan commented:
Thanks all.

@filiph could you please give final confirmation `skeleton` is the name we want to use? 

If so, we'll need to change the `flutter_template_images` package before I can finalize this PR. To speed that process up, could you please help me merge and deploy https://github.com/flutter/packages/pull/310? 

After that, I can update the `flutter_tools` package with the new name and we should be good to go.

filiph commented:
Thanks for waiting! In the absence of strong opinions (other than Chris's and Ian's and mine), `skeleton` wins.

I requested review over at https://github.com/flutter/packages/pull/310.

csells commented:
> Thanks for waiting! In the absence of strong opinions (other than Chris's and Ian's and mine), `skeleton` wins.
> 
> I requested review over at [flutter/packages#310](https://github.com/flutter/packages/pull/310).

Before we settle on skeleton, can we do a more thorough exploration of the naming possibilities? Those first three where just off the top of my head.

csells commented:
Other names to consider include:

Full
Standard
Starter

csells commented:
Android Studio uses more descriptive names like:

No Activity
Basic Activity
Full Screen Activity

Which we could translate into their Flutter equivalents, e.g. Multi-Screen.

csells commented:
VGV's new tool calls their app a "Starter App"

csells commented:
'yo' has thousands with no discernable pattern, so that's not a good source of inspiration

csells commented:
VS is more utilitarian, e.g.

Dynamic-link Library (DLL)
Static Library
Console App
Windows Desktop Application

filiph commented:
I added some of the ideas to a form: https://docs.google.com/forms/d/e/1FAIpQLSc5XrLdqxMKUNHUGROJkncU316VbHHdGInU_VV5LLy2qDrpfQ/viewform?usp=sf_link

Chris, you have edit access to it. I'm OOO today but I thought I'd do at least that.

csells commented:
ASP.NET uses names like:

Core
Core Empty
Web API
gRPC Service (cool!)
Web Application

brianegan commented:
It's good to keep in mind this template adds to the following list of
Flutter templates:

  - app
  - module
  - package
  - plugin

That's why I chose `list_detail_app`. It seems descriptive and fitting for
the current naming structure.

On Fri, Mar 12, 2021 at 6:18 PM Chris Sells ***@***.***>
wrote:

> ASP.NET uses names like:
>
> Core
> Core Empty
> Web API
> gRPC Service (cool!)
> Web Application
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/flutter/pull/76519#issuecomment-797634159>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA65DFXE4B36MG6DSTRIM3TDI5EBANCNFSM4YAM3XIA>
> .
>

csells commented:
I updated @filiph's survey with a couple more options.

@brianegan I see the logic behind an _app suffix but how do we expect them to invoke this new template from the CLI? If they've already chosen -t app, then an the _app suffix seems redundant.

brianegan commented:
To create this template from the CLI, users type in `flutter create -t
list_detal_app .`, they would not use the `-t app` for this one. `-t app`
remains backward compatible as the default Counter app.

On Fri, Mar 12, 2021 at 6:46 PM Chris Sells ***@***.***>
wrote:

> I updated @filiph <https://github.com/filiph>'s survey with a couple more
> options.
>
> @brianegan <https://github.com/brianegan> I see the logic behind an _app
> suffix but how do we expect them to invoke this new template from the CLI?
> If they've already chosen -t app, then an the _app suffix seems redundant.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/flutter/pull/76519#issuecomment-797650634>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA65DGFRJ62DSQYYLNLODTTDJAMVANCNFSM4YAM3XIA>
> .
>

csells commented:
In that case, I can see using a convention like "app" or "package" or "plugin" meaning "default app" etc. and then using longer names like "federated_plugin" or "starter_app" would be a way to expand that namespace w/o stepping on backwards compat.

filiph commented:
@googlebot I consent.

filiph commented:
Hi @Hixie, @zanderso, @jonahwilliams, could you please have a final look and stamp LGTM if all looks good?

The naming decision seems to be `skeleton`, according both to reactions in this thread and my highly scientific quantitative study:

<img width="783" alt="Screen Shot 2021-03-15 at 7 46 55 PM" src="https://user-images.githubusercontent.com/919717/111248570-6d14d500-85c7-11eb-89a9-3659b9f42e69.png">

FWIW, I like `skeleton`, and don't think we can arrive at a significantly better option.

Hixie commented:
I'm fine with skeleton, as noted earlier. :-)

brianegan committed:
Rename list_detail_app to skeleton

brianegan commented:
Thanks everyone. I agree: There's no perfect name here, and I'm happy with skeleton as well :)

I saw some updates were pushed to this branch, but they looked a bit busted. There should only be about ~900 additions, nowhere near 8000. Therefore, I've rebased my branch against master to be sure this PR is up-to-date, renamed everything to skeleton, and force-pushed once again.

At this point, I think the plan for landing this PR seems to be:

  1. Merge `flutter_template_images: 3.0.0`, which renames `list_detail_app` to `skeleton`: https://github.com/flutter/packages/pull/310
  2. Get any final input from the team and update the PR
  3. Rerun the CI checks for this PR and fix any issues. The flutter_tools pubspec for this PR already references flutter_template_images: 3.0.0, but the tests will not pass until that's deployed.
  4. Merge it up!

zanderso commented on packages/flutter_tools/lib/src/flutter_project_metadata.dart:
```diff
@@ -15,6 +15,8 @@ enum FlutterProjectType {
   /// It is different than the "module" template in that it exposes and doesn't
   /// manage the platform code.
   app,
+  /// A List/Detail app template that follows community best practices.
```
nit: it looks like this list was alphabetized.

zanderso reviewed (approved):
lgtm w/ nit

I don't have a strong opinion about the name, but an advantage of 'skeleton_app' would be making it easier to think of the name for similar templates for modules and plugins: 'skeleton_module', 'skeleton_plugin', etc.

goderbauer reviewed (approved):
LGTM

No strong opinion about the particular name. I do see advantages in using the suffix `_app` to make it clear what kind of template this is (e.g. module, plugin, package, or app).

csells commented:
cc @stuartmorgan re: the use of non-default templates using a naming convention like "skeleton_app" or "federated_plugin" as he's particularly interested in adding a federated plugin template in the future.

stuartmorgan commented:
I think it's plausible that we'll need a couple of new `*_plugin` types, but I don't have a strong opinion on whether that means this needs an `*_app` suffix; I don't see an issue with adding `*_plugin` templates even if this doesn't end with `_app`

filiph commented:
The reason I'm not 100% on board with `*_app` is that it's not great developer experience on the command line. `flutter create -t skeleton` is easy to write and remember. `flutter create -t skeleton_app` not so much.

I do understand the appeal of having a more stratified approach to naming. I just think we need to understand the tradeoff between that, and developer experience.

csells commented:
Perhaps we could have a policy that non-app templates get the suffix, e.g. skeleton is an app template but instead of using federated, we use federated_plugin.

Hixie commented:
We could also drop the underscore, as in "skeletonapp"

goderbauer commented:
https://github.com/flutter/flutter/pull/78619 is migrating the other templates to null safety. This one probably also needs to be null save then.

/cc @mit-mit

brianegan commented:
Hey all -- not 100% sure where we stand with this one. Unfortunately, I do not have any more free time to dedicate to this PR. Therefore, I'll close this one out for now.

Please feel free to pick it up if you'd like to continue with it!

Hixie commented:
@brianegan what was left to do? wasn't it reviewed and ready to land?

brianegan commented:
It was ready to roll with the old name or "skeleton," but from what I can gather the naming discussion has not come to a conclusion and the PR has become a bit stale. 

Therefore, to wrap up this work, the team needs to choose a name, update the template to support the new name, update the template to support null safety, and rebase the branch against master again as there are now conflicting changes.

stuartmorgan commented:
If someone is going to pick this up and drive it to completion please let me know soon, because otherwise https://github.com/flutter/packages/pull/295 needs to be reverted as a second breaking change to that package.

csells commented:
In general, this template has a ton of great practices in a very little bit of code. I'm looking forward to using it in my own projects.

Here's a little bit of review feedback for @filiph as referenced to https://github.com/brianegan/new_flutter_template:
- the settings page action is hidden behind the debug banner, which could be mitigated with a tooltip
- I would expect settings to be a dialog on the desktop form factor and not a page to navigate to
- I normally find myself having a shared Scaffold and AppBar between pages to reduce the duplicated code while providing a consistent UI between pages
- the item details page has no Settings actions
- I would expect the strings "Dummy Items" and "Item Details" to be in the arb and for the code to use them
- all of the tests except the first two in implementation_test.dart have great comments; would like great comments in all tests

filiph commented:
Thanks @csells, replies inline:

> the settings page action is hidden behind the debug banner, which could be mitigated with a tooltip

We were aware of the issue but decided to prioritize simplicity of code. What do you mean by mitigating the issue with a tooltip? 

> I would expect settings to be a dialog on the desktop form factor and not a page to navigate to

That would make the code a lot more complex, and isn't necessarily all that clear cut. Desktop apps often open a completely separate window for settings screens, which would make things even harder to implement.

> I normally find myself having a shared Scaffold and AppBar between pages to reduce the duplicated code while providing a consistent UI between pages

We explored this. In many cases, AppBar will be meaningfully different between pages. This can be mitigated by creating a custom widget, of course, but that would make the template more complex. And it's something where we don't need to educate. The template educates about things like file structure, but extracting widget is something an intermediate developer is perfectly capable of.

In short, this goes back to the goals of the template: getting out of the way of an intermediate/advanced developer.

> the item details page has no Settings actions

That is intentional. In most apps we're seeing, app-level settings aren't accessible from everywhere. Of course, a desktop app has its settings in the menu, so in that sense, it's available everywhere. But again, this is something an intermediate developer should be able to implement themselves. And it allows us to save code that would otherwise need to be understood, modified and/or deleted by the developer.

> I would expect the strings "Dummy Items" and "Item Details" to be in the arb and for the code to use them
all of the tests except the first two in implementation_test.dart have great comments; would like great comments in all tests 

This is also intentional. We've seen that developers, when starting projects, don't immediately implement i18n everywhere. It really damages perceived developer productivity when, instead of `Text('OK')` you have to write the i18n equivalent (in several files). 

In the template, we show how to go about doing i18n in Flutter, and give you the files in the right places and with the right contents, but don't implement it fully for you. As a developer, you'd be deleting most of it anyway.

filiph commented:
ping @csells

csells commented:
@filiph replies:

re: mitigating with a tooltip: at least a tooltip would show if you move the mouse over in that direction, although obviously that's a pretty lame mitigation.

I'm curious what feedback we get from a wider audience on the rest. Certainly what you have is a big improvement in pointing people down the right path structurally compared to the Counter app.

Hixie commented:
Is there an updated PR to track work on this?

filiph commented:
Not yet. I'm planning to create it after next week (I/O).

christopherfujino commented:
> Not yet. I'm planning to create it after next week (I/O).

FYI @filiph the flutter/packages change got reverted https://github.com/flutter/packages/pull/353. You'll need to re-land that in order to land the tool change.

stuartmorgan commented:
> You'll need to re-land that in order to land the tool change.

Given the experience last time, where the images package was left in an unusable state for several months, I would strongly prefer that the updated PR be *fully approved and ready to land* before doing the changes it the images package. It can be tested locally with version 2.0.0 of the package.

filiph commented:
Hey all, I'm unable to push changes to this PR for whatever reason, so I'm recreating it in https://github.com/flutter/flutter/pull/83530. That PR is ready for review. I'll ask folks over there.

This PR can stay closed.
